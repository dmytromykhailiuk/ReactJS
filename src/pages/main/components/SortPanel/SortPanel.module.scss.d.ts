// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'sort-panel': string;
  'sort-panel__arrow-down': string;
  'sort-panel__arrow-up': string;
  'sort-panel__option': string;
  'sort-panel__options': string;
  'sort-panel__title': string;
}
export const cssExports: CssExports;
export default cssExports;
