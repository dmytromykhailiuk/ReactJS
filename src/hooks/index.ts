export * from "./useInputValue";
export * from "./useLocalTime";
