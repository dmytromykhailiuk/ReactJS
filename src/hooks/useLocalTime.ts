import { useState, useEffect } from "react";

const getLocalTimeString = () => new Date().toLocaleTimeString("it-IT");

export function useLocalTime(delay = 1000) {
  const [localTime, changeLocalTime] = useState(getLocalTimeString());

  useEffect(() => {
    const interval = setInterval(() => {
      changeLocalTime(getLocalTimeString());
    }, delay);

    return () => {
      clearInterval(interval);
    };
  }, []);

  return localTime;
}
