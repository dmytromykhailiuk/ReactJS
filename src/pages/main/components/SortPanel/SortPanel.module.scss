@import "../../../../scss/variables";
@import "../../../../scss/mixins.scss";

.sort-panel {
  margin: 0;
  color: $white-color;
  display: flex;
  margin-top: 40px;

  @include medium-screen {
    margin-top: 19px;
  }

  &__title {
    font-size: 18px;
    margin-right: 30px;
  }

  &__options {
    list-style: none;
  }

  &__option {
    font-size: 18px;
    cursor: pointer;
    padding-bottom: 20px;
    display: flex;

    &:hover {
      color: $pink-color;
      transition: 0.4s;
    }
  }

  &__arrow-down, &__arrow-up {
    width: 0; 
    height: 0;
    border-left: 5px solid transparent;
    border-right: 5px solid transparent;
    margin-top: 6px;
    margin-left: 12px;
  }

  &__arrow-down {
    border-top: 7px solid $pink-color;
  }

  &__arrow-up {
    border-bottom: 7px solid $pink-color;
  }
}
