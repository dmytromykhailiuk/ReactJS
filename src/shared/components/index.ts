export * from "./Footer";
export * from "./Logo";
export * from "./Button";
export * from "./Input";
export * from "./ThreeDotsIcon";
export * from "./ErrorBoundary";
