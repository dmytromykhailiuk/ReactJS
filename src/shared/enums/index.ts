export * from "./button-types";
export * from "./button-size";
export * from "./categories";
export * from "./sorting-options";
export * from "./movie-form-values";
export * from "./router-paths";
export * from "./modal-types";
