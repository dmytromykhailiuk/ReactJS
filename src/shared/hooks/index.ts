export * from "./useInputValue";
export * from "./useClickOutside";
