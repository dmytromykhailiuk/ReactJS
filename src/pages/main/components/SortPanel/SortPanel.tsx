import React, { useCallback, useEffect, useMemo, useState } from 'react';
import classes from './SortPanel.module.scss';
import { SortingOptions, SortingOptionsProperties } from 'shared/enums';
import { clickOutside } from 'shared/helpers';

interface SortPanelProps {
  onChangeSortingDirection?: (isDownDirection: boolean) => void;
  setSortingOption: (sortingOption: SortingOptionsProperties) => void;
  isDownDirection?: boolean;
  sortingOption: SortingOptionsProperties;
}

const sortingOptions: SortingOptionsProperties[] = [SortingOptionsProperties.RELEASE_DATE, SortingOptionsProperties.TITLE];

const SortPanel: React.FC<SortPanelProps> = ({ onChangeSortingDirection = () => {}, isDownDirection = true, sortingOption, setSortingOption }) => {
  const [shouldShowOptions, setShouldShowOptionsValue] = useState(false);

  const triggerSetShouldShowOptionsValue = useCallback(() => {
    setShouldShowOptionsValue(!shouldShowOptions);
  }, [shouldShowOptions]);

  const hideOptions = useCallback(() => {
    setShouldShowOptionsValue(false);
  }, []);

  const onSetSortingOption = useCallback((sortingOption, shouldShowOptions) => {
    setShouldShowOptionsValue(!shouldShowOptions);
    setSortingOption(sortingOption);
  }, []);

  const clickOutsides = useMemo(() => clickOutside(classes['sort-panel__options-wrapper'], hideOptions), [])

  useEffect(() => {
    if (shouldShowOptions) {
      document.addEventListener("click", clickOutsides);
      return () => {
        document.removeEventListener("click", clickOutsides);
      }
    }
  }, [shouldShowOptions])

  const onChangeDirection = useCallback(() => {
    onChangeSortingDirection(!isDownDirection);
  }, [isDownDirection]);

  return (
    <div className={classes['sort-panel']}>
      <div className={classes['sort-panel__options-wrapper']}>
        <div className={classes['sort-panel__title']} onClick={triggerSetShouldShowOptionsValue}>SORT BY</div>
        { shouldShowOptions && <div className={classes['sort-panel__options']}>
          { sortingOptions.map(sortingOption => (
            <div 
              className={classes['sort-panel__option']}
              key={sortingOption}
              onClick={() => onSetSortingOption(sortingOption, shouldShowOptions)}
            >
              { SortingOptions[sortingOption] }
            </div>
          )) }
        </div>}
      </div>
      <ul 
        className={classes['sort-panel__selected-option-wrapper']}
        onClick={onChangeDirection}
      >
        <li className={classes['sort-panel__selected-option']}>
          { SortingOptions[sortingOption] }
          <div className={
            classes[isDownDirection ? "sort-panel__arrow-down" : "sort-panel__arrow-up"]
          }></div>
        </li>
      </ul>
    </div>
  );
}

export default SortPanel;
