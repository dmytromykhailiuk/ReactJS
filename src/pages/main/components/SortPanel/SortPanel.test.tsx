import { SortingOptionsProperties } from 'shared/enums';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import toJson from 'enzyme-to-json';
import { useClickOutside } from 'shared/hooks';
import SortPanel from './SortPanel';

jest.mock('shared/hooks');

let shouldCallClickOusideFn = true;

(useClickOutside as jest.Mock).mockImplementation((_, fn) => {
  if (shouldCallClickOusideFn) {
    shouldCallClickOusideFn = false;
    fn();
  }
});

describe('SortPanel', () => {
  let wrapper: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

  const onChangeSortingDirection = jest.fn();
  const setSortingOption = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    wrapper = shallow(
      <SortPanel
        isDownDirection
        sortingOption={SortingOptionsProperties.DURATION}
        onChangeSortingDirection={onChangeSortingDirection}
        setSortingOption={setSortingOption}
      />,
    );
  });
  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should show 'arrow-up' when isDownDirection equal false", () => {
    wrapper.setProps({
      isDownDirection: false,
    });
    expect(wrapper.exists('.sort-panel__arrow-up')).toBeTruthy();
  });

  it("should show sorting options ater click on 'sort-panel__title'", () => {
    wrapper.find('.sort-panel__title').simulate('click');
    expect(wrapper.exists('.sort-panel__option')).toBeTruthy();
  });

  it("should call onChangeSortingDirection when click on 'sort-panel__selected-option-wrapper'", () => {
    wrapper.find('.sort-panel__selected-option-wrapper').simulate('click');
    expect(onChangeSortingDirection).toHaveBeenCalled();
  });

  it("should call setSortingOption when click on 'sort-panel__option'", () => {
    wrapper.find('.sort-panel__title').simulate('click');
    wrapper.find('.sort-panel__option').first().simulate('click');
    expect(setSortingOption).toHaveBeenCalled();
  });
});
