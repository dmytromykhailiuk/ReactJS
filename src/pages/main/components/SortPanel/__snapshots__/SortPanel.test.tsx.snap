// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortPanel should match snapshot 1`] = `
<div
  className="sort-panel"
>
  <div
    className="sort-panel__options-wrapper"
  >
    <div
      className="sort-panel__title"
      onClick={[Function]}
    >
      SORT BY
    </div>
  </div>
  <ul
    className="sort-panel__selected-option-wrapper"
    onClick={[Function]}
  >
    <li
      className="sort-panel__selected-option"
    >
      DURATION
      <div
        className="sort-panel__arrow-down"
      />
    </li>
  </ul>
</div>
`;
